Add async support

skeggse opened the pull request:
Also brings `master` up to date with upstream.

TODO:

- [x] ~Test all built-in expressions for async support~

icedtoast committed:
Fix $eq & $ne to work with complex objects

Call comparable on both sides, so that the correct logic is followed
when filtering with $eq or $ne.

omanizer committed:
Bring to mongo spec for $ne on fields of arrays of objects.  If running a $ne operation on an array of sub-objects, the object being filtered should only pass if all of the sub-objects pass the $ne query.  Prior to this the object would pass if only one of the sub-objects passed the $ne query.

Commit:
Merge pull request #144 from omanizer/master

Bring to mongo spec for $ne on fields of arrays of objects.

crcn committed:
major bump to 6.0.0 for behavior change

crcn committed:
Update README.md

j2L4e committed:
Update README.md

fix readme typo

Commit:
Merge pull request #145 from j2L4e/patch-1

Update README.md

Commit:
Merge pull request #141 from SmileSoftware/master

Fix $eq & $ne to work with complex objects

crcn committed:
refactor to es6 & clean up

crcn committed:
add es6 path

crcn committed:
upgrade node

crcn committed:
update test command

crcn committed:
check exactness of object

crcn committed:
update changelog

crcn committed:
update lib

crcn committed:
fix lib

crcn committed:
fix conflict

crcn committed:
start on async tests

crcn committed:
added test for  matching

crcn committed:
ability to define custom expressions

crcn committed:
add docs

crcn committed:
remove deprecated tests

crcn committed:
Update README.md

Commit:
Add async support

Also rewrites lots of things because prettier, eslint etc.

ghmeier commented on src/index.js:
```diff
@@ -126,169 +108,161 @@ var expressions = {
-          if (a[i] == null) {
-            return true;
-          }
+      return false;
```
Yay for early returns!

ghmeier commented on src/index.js:
```diff
@@ -126,169 +108,161 @@ var expressions = {
+    /*
+      Handles the case of {'field': {$in: [/regexp1/, /regexp2/, ...]}}
+    */
+    return maybeAsyncThen(
```
So this is 'maybe run the internal filter as async and then return the result of `a.indexOf(comparableB) >= 0`'?

skeggse replied:
Almost - it mirrors the prior code where it only considers `a.indexOf(comparableB) >= 0` if the possibly-async validators come up `false`.

ghmeier commented on src/index.js:
```diff
@@ -305,113 +279,103 @@ var prepare = {
-  $all: function(a) {
-    return prepare.$and(a);
+  $all(a, query, parseSub) {
+    return prepare.$and(a, query, parseSub);
```
Oh interesting, where did these extra parameters come from?

skeggse replied:
They're used in `$and` and always passed in from the `parse` function - this was apparently a preexisting bug because it started failing tests when I tweaked a few things.

ghmeier replied:
oooh

ghmeier commented on src/index.js:
```diff
@@ -305,113 +279,103 @@ var prepare = {
 
-  $regex: function(a, query) {
-    return new RegExp(a, query.$options);
+  $regex(regexParam, query) {
```
thanks for the rename!!!

ghmeier commented on src/index.js:
```diff
@@ -447,8 +411,11 @@ function findValues(current, keypath, index, object, values) {
-function createNestedValidator(keypath, a, q) {
-  return { a: { k: keypath, nv: a, q: q }, v: nestedValidator };
+function createNestedValidator(keypath, validator, query) {
+  const arg = {path: keypath, validator, query};
```
Just checking - is this the only place `nestedValidator` is called and so the only place we need to change the values of `arg`?

skeggse replied:
Yes!

ghmeier commented on src/index.js:
```diff
@@ -460,50 +427,47 @@ function isVanillaObject(value) {
-        (options && options.expressions && options.expressions[key]);
+        (options &&
+          options.expressions &&
+          has(options.expressions, key) &&
```
nice extra safety check!

skeggse replied:
:grin:

ghmeier commented on test/basic-test.js:
```diff
@@ -1,11 +1,11 @@
-import * as assert from "assert";
-import sift, { indexOf as siftIndexOf } from "..";
+import * as assert from 'assert';
```
Just prettier changes here?

skeggse replied:
Yup. I should've separated out the prettier changes but by the time I realized that I'd already committed :|

ghmeier commented on test/custom-operations-test.js:
```diff
@@ -1,24 +1,21 @@
-  it("can add a custom $mod operation", function() {
-    var filter = sift(
-      { $mod2: 2 },
+describe('custom operations', () => {
```
Could we add a custom async filter test?

skeggse replied:
0c6aee53d8d9126daa5f9cc9add9612fc80c6943

ghmeier reviewed (approved):
Nice! Just a couple questions and a test :)

shils commented:
We should verify that we're not affected by the behavioral change mentioned here - https://github.com/crcn/sift.js/issues/117.

Commit:
Remove vulnerable $where implementation

Commit:
Use jest instead of mocha

Commit:
Prefer object shorthand

Commit:
Remove unused vars

Commit:
Extract asyncFilter to utils

Commit:
Test custom async operations

Commit:
Test async utilities

Commit:
Test more edge cases

This increases code coverage

Commit:
Improve $type support

Commit:
Remove exports warning

Commit:
Remove unnecessary build

The tests now reference the prebuilt files directly so this isn't needed.

Commit:
Document deviant $size behavior

MongoDB doesn't implement $size this way, at least as far as their
documentation is concerned. Call this out explicitly, and we can
consider whether it warrants a major bump to alter this.

Commit:
Move RegExp checking to a utility function

Commit:
Fix maybeAsyncSome implementation

Commit:
Rename bindings for readability

Commit:
Remove unnecessary async handling

We don't actually support asynchronous work inside these expressions so
no point in shelling out to maybeAsyncSome/maybeAsyncEvery.

Commit:
Fix typo

Commit:
Document use of indexOf over includes

Commit:
Check $regex types

Commit:
Avoid repeat work

Commit:
Test more custom async operations

Commit:
Make wording consistent

Commit:
Fix buffer allocation

shils reviewed (approved):
LGTM once all built-in expressions are tested for async support.

skeggse commented:
>LGTM once all built-in expressions are tested for async support.

This is complete - in `custom-operations.test.js`, as those tests depend on actually having async operations available. Not all built-in operations actually support async operations. For example, `$size` doesn't accept an expression as its parameter - it expects a constant number and thus supporting async ops doesn't impact that operator. Sound good?

shils commented:
@skeggse yup! My comment was just in reference to the TODO which hadn't been checked off at the time of my comment.

Commit:
Hook build into publish flow
